### PR TITLE
Integrate hardfork in genesis process

### PR DESCRIPTION
**Pull request by sasurobert**

### CODE REVIEW

File: genesis/process/genesisBlockCreator.go [hunk 1]
```diff
@@ -29,11 +34,58 @@ func NewGenesisBlockCreator(arg ArgsGenesisBlockCreator) (*genesisBlockCreator,
...
-	return &genesisBlockCreator{
+	gbc := &genesisBlockCreator{
 		arg:                 arg,
 		shardCreatorHandler: CreateShardGenesisBlock,
 		metaCreatorHandler:  CreateMetaGenesisBlock,
-	}, nil
+	}
+
+	if arg.HardForkConfig.MustImport {
+		err := gbc.createHardForkImportHandler()
```
Comment (iulianpascalau) [inline genesis/process/genesisBlockCreator.go:44]: we should think of a way to invalidate the must import flag after the first successful sync of the network. Might use a max nonce until this flag is validated?

File: genesis/process/metaGenesisBlockCreator.go [hunk 2]
```diff
@@ -65,6 +66,15 @@ func CreateMetaGenesisBlock(arg ArgsGenesisBlockCreator) (data.HeaderHandler, er
 		return nil, err
 	}
 
+	if arg.HardForkConfig.MustImport {
+		//TODO: think how to integrate when genesis is modified as well - how to integrate with imported data
+		// one example being - delete or not the validator statistics trie - and restart from 0, or should import from old
```
Comment (iulianpascalau) [inline genesis/process/metaGenesisBlockCreator.go:71]: Answer & discussed: recreate the validator statistics trie from the (newly) nodesSetup.json. Or a new predictable nodesSetup_1.json or something like this.

File: genesis/process/shardGenesisBlockCreator.go [hunk 3]
```diff
@@ -36,6 +37,11 @@ func CreateShardGenesisBlock(arg ArgsGenesisBlockCreator) (data.HeaderHandler, e
 	}
 
 	//TODO add here delegation process
+	if arg.HardForkConfig.MustImport {
+		//TODO think about how to integrate when genesis is modified as well for hardfork - when should set balances be called
```
Comment (iulianpascalau) [inline genesis/process/shardGenesisBlockCreator.go:41]: Is a must to keep the old, original genesis block? No one will care about it except the entities that want to sync from the start.

File: process/block/baseProcess.go [hunk 4]
```diff
@@ -118,7 +119,7 @@ func (bp *baseProcessor) checkBlockValidity(
 	currentBlockHeader := bp.blockChain.GetCurrentBlockHeader()
 
 	if check.IfNil(currentBlockHeader) {
-		if headerHandler.GetNonce() == 1 { // first block after genesis
+		if headerHandler.GetNonce() == bp.genesisNonce+1 { // first block after genesis
```
Comment (iulianpascalau) [inline process/block/baseProcess.go:122]: :+1:

File: genesis/process/genesisBlockCreator.go (hunk 1 above)
Comment (sasurobert) [reply genesis/process/genesisBlockCreator.go:44]: after a hardfork everybody has to start and create the genesis block through this method - so it is ok. MustImport as a flag will be set only when there is a hardfork - for that you need to set export and import storage / folder.

File: genesis/process/shardGenesisBlockCreator.go (hunk 3 above)
Comment (sasurobert) [reply genesis/process/shardGenesisBlockCreator.go:41]: agreed. at this moment it we do not care about initial genesis.

File: genesis/process/shardGenesisBlockCreator.go [hunk 5]
```diff
@@ -53,6 +59,59 @@ func CreateShardGenesisBlock(arg ArgsGenesisBlockCreator) (data.HeaderHandler, e
...
+		arg.ChainID,
+		arg.HardForkConfig.StartRound,
+		arg.HardForkConfig.StartNonce,
+		arg.HardForkConfig.StartEpoch,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	errNotCritical := processors.txCoordinator.SaveBlockDataToStorage(bodyHandler.(*block.Body))
```
Comment (SebastianMarian) [inline genesis/process/shardGenesisBlockCreator.go:104]: I would extract bodyHandler.(*block.Body) into variable and check it against type assertion before saving it into storage

File: process/block/metablock.go [hunk 6]
```diff
@@ -1626,8 +1641,11 @@ func (mp *metaProcessor) computeExistingAndRequestMissingShardHeaders(metaBlock
 	mp.hdrsForCurrBlock.mutHdrsForBlock.Lock()
 	defer mp.hdrsForCurrBlock.mutHdrsForBlock.Unlock()
 
-	for i := 0; i < len(metaBlock.ShardInfo); i++ {
-		shardData := metaBlock.ShardInfo[i]
+	for _, shardData := range metaBlock.ShardInfo {
+		if shardData.Nonce == 0 {
+			continue
```
Comment (SebastianMarian) [inline process/block/metablock.go:1646]: Do we not need to execute lines 1163-1670 before continue? Maybe later we need to read from mp.hdrsForCurrBlock.hdrHashAndInfo the genesis shard header(to save it or other stuff) , or we should add this skip everywhere.

File: process/peer/process.go [hunk 7]
```diff
@@ -690,6 +693,10 @@ func (vs *validatorStatistics) updateShardDataPeerState(header data.HeaderHandle
 	}
 
 	for _, h := range metaHeader.ShardInfo {
+		if h.Nonce == 0 {
```
Comment (SebastianMarian) [inline process/peer/process.go:696]: == vs.genesisNonce

File: process/block/metablock.go [hunk 8]
```diff
@@ -1626,8 +1641,11 @@ func (mp *metaProcessor) computeExistingAndRequestMissingShardHeaders(metaBlock
 	mp.hdrsForCurrBlock.mutHdrsForBlock.Lock()
 	defer mp.hdrsForCurrBlock.mutHdrsForBlock.Unlock()
 
-	for i := 0; i < len(metaBlock.ShardInfo); i++ {
-		shardData := metaBlock.ShardInfo[i]
+	for _, shardData := range metaBlock.ShardInfo {
+		if shardData.Nonce == 0 {
```
Comment (SebastianMarian) [inline process/block/metablock.go:1645]: == mp.genesisNonce

File: genesis/process/metaGenesisBlockCreator.go [hunk 9]
```diff
@@ -95,6 +105,47 @@ func CreateMetaGenesisBlock(arg ArgsGenesisBlockCreator) (data.HeaderHandler, er
...
+	metaHdr.SetTimeStamp(arg.GenesisTime)
+	metaHdr.EpochStart.Economics = block.Economics{
+		TotalSupply:            big.NewInt(0).Set(arg.Economics.GenesisTotalSupply()),
+		TotalToDistribute:      big.NewInt(0),
+		TotalNewlyMinted:       big.NewInt(0),
+		RewardsPerBlockPerNode: big.NewInt(0),
+		NodePrice:              big.NewInt(0).Set(arg.Economics.GenesisNodePrice()),
+	}
+
+	errNotCritical := processors.txCoordinator.SaveBlockDataToStorage(bodyHandler.(*block.Body))
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply genesis/process/metaGenesisBlockCreator.go:141]: done

File: genesis/process/shardGenesisBlockCreator.go (hunk 5 above)
Comment (sasurobert) [reply genesis/process/shardGenesisBlockCreator.go:104]: done

File: process/peer/process.go (hunk 7 above)
Comment (sasurobert) [reply process/peer/process.go:696]: good catch.

File: process/block/metablock.go (hunk 8 above)
Comment (sasurobert) [reply process/block/metablock.go:1645]: good catch.

File: process/block/metablock.go (hunk 6 above)
Comment (sasurobert) [reply process/block/metablock.go:1646]: putting to hdrsForCurrBlock is done at checkShardHeadersValidity

File: genesis/process/metaGenesisBlockCreator.go (hunk 2 above)
Comment (sasurobert) [reply genesis/process/metaGenesisBlockCreator.go:71]: would leave like this now, we can decide about this later.

**Pull request merged**